### ISSUE TITLE
ProGuard strips out GoogleJsonError$ErrorInfo

### ISSUE THREAD

**Issue by wonderfly** [googleapis/google-api-java-client#542]
_From [yan...@google.com](https://code.google.com/u/111815493751521097215/) on June 18, 2012 09:26:08_

Version of google-api-java-client (e.g. 1.5.0-beta)? 1.10.2-beta Java environment (e.g. Java 6, Android 2.3, App Engine)? Android Describe the problem. While parsing an error at runtime on a ProGuarded calendar-android-sample, I got this serious exception:

06-18 12:09:36.419: E/AndroidRuntime(9172): Caused by: java.lang.TypeNotPresentException: Type com.google.api.client.googleapis.json.GoogleJsonError$ErrorInfo not present

The problem is that ProGuard strips out GoogleJsonError$ErrorInfo because it thinks it is unused.  If you look at the source, these are the usages:

  @Key
  private List<ErrorInfo> errors;
  public final List<ErrorInfo> getErrors() {
    return errors;
  }
  public final void setErrors(List<ErrorInfo> errors) {
    this.errors = errors;
  }

In other words, it is only used as a generic type parameter, but this information is stripped out before ProGuard sees it.

This is actually a more general problem, and for now we will use an ad-hoc process to try to find these problems in our libraries.

The work-around is to add this to the proguard configuration file:

-keep class com.google.api.client.googleapis.json.GoogleJsonError$ErrorInfo

but many developers won't know they need to add it. How would you expect it to be fixed? A simple work-around can be to do something like this in GoogleJsonError:

  static {
    Data.nullOf(ErrorInfo.class);
  }

Not pretty, but it works.

_Original issue: http://code.google.com/p/google-api-java-client/issues/detail?id=527_
